hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/file/FileFormat.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.pipeline.file;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.nio.charset.Charset;
+
+/**
+ * Identifies the data format of a file to be used as a Jet data source.
+ * This is a data object that holds the configuration; actual implementation
+ * code is looked up elsewhere, by using this object as a key.
+ *
+ * @param <T> type of items a source using this file format will emit
+ * @since 4.4
+ */
+public interface FileFormat<T> {
+
+    /**
+     * Returns the unique identifier of the file format. The convention is to
+     * use the well-known filename suffix or, if there is none, a short-form
+     * name of the format.
+     */
+    @Nonnull
+    String format();
+
+
+    // Factory methods for supported file formats are here for easy discoverability.
+
+    /**
+     * Returns a file format for Avro files.
+     */
+    @Nonnull
+    static <T> AvroFileFormat<T> avro() {
+        return avro(null);
+    }
+
+    /**
+     * Returns a file format for Avro files that specifies to use reflection
+     * to deserialize the data into instances of the provided Java class.
+     * Jet will use the {@code ReflectDatumReader} to read Avro data. The
+     * parameter may be {@code null}, disabling the option to deserialize
+     * using reflection, but for that case you should prefer the no-argument
+     * {@link #avro()} call.
+     */
+    @Nonnull
+    static <T> AvroFileFormat<T> avro(@Nullable Class<T> clazz) {
```

Comment:
why don't we use `Nonnull` and force user to use the no-argument variant. We use this convention for `LinesTextFileFormat` and `TextFileFormat` for example.

Reply:
See comment on the related method.